### PR TITLE
[SPARK-11243] [SQL] output UnsafeRow from columnar cache

### PR DESCRIPTION
**Pull request by davies**
This PR change InMemoryTableScan to output UnsafeRow, and optimize the unrolling and scanning by coping the bytes for var-length types between UnsafeRow and ByteBuffer directly without creating the wrapper objects. When scanning the decimals in TPC-DS store_sales table, it's 80% faster (copy it as long without create Decimal objects).

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 1]
```diff
@@ -95,41 +99,75 @@ public void alignToWords(int numBytes) {
...
+    if (Double.isNaN(value)) {
+      value = Double.NaN;
+    }
+    Platform.putDouble(holder.buffer, getFieldOffset(ordinal), value);
+  }
+
+  public void write(int ordinal, Decimal input, int precision, int scale) {
+    if (precision <= Decimal.MAX_LONG_DIGITS()) {
+      // make sure Decimal object has the same scale as DecimalType
+      if (input.changePrecision(precision, scale)) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: as discussed offline, maybe change this to assert?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 2]
```diff
@@ -618,6 +618,25 @@ public void writeTo(ByteBuffer buffer) {
     buffer.position(pos + sizeInBytes);
   }
 
+  /**
+   * Write the bytes of var-length field into ByteBuffer
+   */
+  public void writeFieldTo(int ordinal, ByteBuffer buffer) {
+    final long offsetAndSize = getLong(ordinal);
+    final int offset = (int) (offsetAndSize >> 32);
+    final int size = (int) (offsetAndSize & ((1L << 32) - 1));
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: can we create a constant for (1L << 32) - 1?

We have a lot of them now in this file. LOWER_32_BIT_MASK?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 3]
```diff
@@ -618,6 +618,25 @@ public void writeTo(ByteBuffer buffer) {
     buffer.position(pos + sizeInBytes);
   }
 
+  /**
+   * Write the bytes of var-length field into ByteBuffer
+   */
+  public void writeFieldTo(int ordinal, ByteBuffer buffer) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: should document this only supports HeapByteBuffer

**Comment by JoshRosen** [comment]
I'm really excited to see how much of a performance difference this makes when scanning string columns, since this could potentially provide a big perf. boost to the sc.textFile replacement that I showed the other day.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 4]
```diff
@@ -58,6 +58,10 @@ private void zeroOutPaddingBytes(int numBytes) {
     }
   }
 
+  public boolean isNullAt(int ordinal) {
+    return BitSetMethods.isSet(holder.buffer, startingOffset, ordinal);
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Do we ever start from the middle of an existing buffer? If not, can we move startingOffset into holder class?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 5]
```diff
@@ -95,41 +99,75 @@ public void alignToWords(int numBytes) {
...
+      // grow the global buffer before writing data.
+      holder.grow(16);
+
+      // zero-out the bytes
+      Platform.putLong(holder.buffer, holder.cursor, 0L);
+      Platform.putLong(holder.buffer, holder.cursor + 8, 0L);
+
+      // Make sure Decimal object has the same scale as DecimalType.
+      // Note that we may pass in null Decimal object to set null for it.
+      if (input == null || !input.changePrecision(precision, scale)) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: assert the 2nd part?

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala [hunk 6]
```diff
@@ -20,18 +20,42 @@ package org.apache.spark.sql.columnar
...
+ * An Iterator to walk through the InternalRows from a CachedBatch
  */
 abstract class ColumnarIterator extends Iterator[InternalRow] {
-  def initialize(input: Iterator[CachedBatch], mutableRow: MutableRow, columnTypes: Array[DataType],
+  def initialize(input: Iterator[CachedBatch], columnTypes: Array[DataType],
     columnIndexes: Array[Int]): Unit
 }
 
 /**
+ * An helper class to update the fields of UnsafeRow, used by ColumnAccessor
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: should put a giant warning here that writes must be in increasing order of ordinals

**Comment by rxin** [comment]
LGTM otherwise.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Yes, in UnsafeProjection, there could be multiple UnsafeRowWriter share a single BufferHolder.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 5 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Will do in separate PR

**Comment by davies** [comment]
@JoshRosen If we only scan the cached string and access it (without adding a ConvertToUnsafe), this could be a little bit slower, because we copy the bytes into UnsafeRow now. For the UTF8String object, we still will create it anyway, before this patch, we create it in ColumnAccessor, after this patch, we create in UnsafeRow.getUTF8String() (we could re-use that actually).

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Actually, there are not needed, `(int) offsetAndSize` should be enough

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: will do in separate PR

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java [hunk 7]
```diff
@@ -64,29 +63,72 @@ public void setOffset(int ordinal) {
...
 
-  public void writeCompactDecimal(int ordinal, Decimal input, int precision, int scale) {
-    // make sure Decimal object has the same scale as DecimalType
-    if (input.changePrecision(precision, scale)) {
-      Platform.putLong(holder.buffer, holder.cursor, input.toUnscaledLong());
-      setOffset(ordinal);
-      holder.cursor += 8;
-    } else {
-      setNullAt(ordinal);
+  public void write(int ordinal, boolean value) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java]: Why implement these primitive write method in writer class not in codegen?

Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java]: Those in UnsafeRowWriter are needed for ColumnAccessor. The purpose of these *Writer is to simplify generated code (which is already hard to understand), these method also go in this direction.

Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java]: I think it is simpler to put them there rather than in generated code.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 8]
```diff
@@ -58,6 +58,10 @@ private void zeroOutPaddingBytes(int numBytes) {
     }
   }
 
+  public boolean isNullAt(int ordinal) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: where do we use this method?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 9]
```diff
@@ -95,41 +99,75 @@ public void alignToWords(int numBytes) {
...
 
-  public void writeCompactDecimal(int ordinal, Decimal input, int precision, int scale) {
-    // make sure Decimal object has the same scale as DecimalType
-    if (input.changePrecision(precision, scale)) {
-      Platform.putLong(holder.buffer, getFieldOffset(ordinal), input.toUnscaledLong());
-    } else {
-      setNullAt(ordinal);
-    }
+  public void write(int ordinal, boolean value) {
+    Platform.putBoolean(holder.buffer, getFieldOffset(ordinal), value);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: I think we should zero-out the whole 8 bytes before writing.
For inner unsafe row, the start position of it may change during mulitly wirtes. So these 8 bytes are not always clean, if we only write a boolean here, the remaing 7 bytes maybe not zero, and will break the equality of unsafe row.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 10]
```diff
@@ -124,17 +124,10 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
           case _ if ctx.isPrimitiveType(dt) =>
-            val fieldOffset = ctx.freshName("fieldOffset")
             s"""
-              final long $fieldOffset = $rowWriter.getFieldOffset($index);
-              Platform.putLong($bufferHolder.buffer, $fieldOffset, 0L);
-              ${writePrimitiveType(ctx, input.value, dt, s"$bufferHolder.buffer", fieldOffset)}
+              $rowWriter.write($index, ${input.value});
             """
 
-          case t: DecimalType if t.precision <= Decimal.MAX_LONG_DIGITS =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: I think a benefit of this is removing a if branch as we know the decimal type before generating code. Will this affect performance?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: it won't matter as much, since the branch predictor can mostly eliminate them.

**Comment by rxin** [comment]
I'm going to merge this first. @davies please take a look at @cloud-fan 's comments, and address them in follow-up prs if necessary.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 9 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Thanks, will fix it later.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 8 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: This is needed by larger Decimal, to call write(decimal) after setNullAt()
